### ISSUE TITLE
Unmanaged windows fail on AMDGPU

### ISSUE THREAD

**Issue by wacossusca34** [wacossusca34/glava#79]
I accidentally encountered this issue on my system while swapping around my multiple GPUs in order to get virtualization setup. I suspect this relates to the AMDGPU Xorg driver, as I was not able to encounter it via Xephyr & llvmpipe.

If anyone _else_ could confirm the problem on a system with `amdgpu` and the following option in `rc.glsl`:

```GLSL
#request setxwintype "!-"
```

Ideally it should display a window just above the root window, and behind everything else.

**Comment by wacossusca34** [comment]
This may be a bug in the `amdgpu` Xorg driver. `xwininfo -id <glava window id>` prints:

```
xwininfo: Window id: 0x2000002 "GLava"

  Absolute upper-left X:  17
  Absolute upper-left Y:  1722
  Relative upper-left X:  17
  Relative upper-left Y:  1722
  Width: 1886
  Height: 380
  Depth: 24
  Visual: 0x4e1
  Visual Class: TrueColor
  Border width: 0
  Class: InputOutput
  Colormap: 0x2000001 (not installed)
  Bit Gravity State: ForgetGravity
  Window Gravity State: NorthWestGravity
  Backing Store State: NotUseful
  Save Under State: no
  Map State: IsViewable                                        <-- !!
  Override Redirect State: yes                                 <-- !!
  Corners:  +17+1722  -17+1722  -17--1022  +17--1022
  -geometry 1886x380+17+1722
 ```

The window is clearly mapped/visible according to the server, yet GLava will not draw. I'm guessing Xorg is reporting the window to be obscured - relevant source (`render.c`):

```C
/* event processing */

            case VisibilityNotify:
                switch (ev.xvisibility.state) {
                    case VisibilityFullyObscured:
                        w->should_render = false;
                        break;
                    case VisibilityUnobscured:
                    case VisibilityPartiallyObscured:
                        w->should_render = true;
                        break;
                    default:
                        fprintf(stderr, "Invalid VisibilityNotify event state (%d)\n", ev.xvisibility.state);
                        break;
                }
                break;

/* ... */

static bool should_render(struct glxwin* w) {
    /* For nearly all window managers, windows are 'minimized' by unmapping parent windows.
       VisibilityNotify events are not sent in these instances, so we have to read window
       attributes to see if our window isn't viewable. */
    XWindowAttributes attrs;
    XGetWindowAttributes(display, w->w, &attrs);
    process_events(w);
    return w->should_render && attrs.map_state == IsViewable;
}
```

Some things of note:

* Removing the `should_render` checks causes `XShmGetImage` to fail. Using `"native"`/`"none"` opacity causes the render loop to run, but there is no visible window.
* `xdotool windowfocus $(xdotool search --class GLava)` is able to grab input focus, does not make the window visible.
* Raising the window does nothing (same as using `!+`).
* A single `VisibilityFullyObscured` event is sent to the window, no subsequent visibility events sent.

**Simply put: `override_redirect`/`CWOverrideRedirect` windows are broken on `amdgpu`.**

**Comment by wacossusca34** [comment]
This is unfortunately something I have little control over. `override_redirect` is also generally considered an evil trick; very few applications make use of this flag aside from some very old Xorg apps (ab)using this for fullscreen. It is, however, a documented feature that should work, so I will get around to reporting this bug later.

**Comment by wacossusca34** [comment]
Fixed in the Xorg driver.
